nabs/suspenders.py

```diff
@@ -0,0 +1,179 @@
+import asyncio
+import copy
+import logging
+from threading import RLock, Event
+
+from bluesky.plan_stubs import null, wait_for
+from bluesky.preprocessors import plan_mutator
+
+logger = logging.getLogger(__name__)
+
+
+class SuspendPreprocessor:
+    """
+    Base preprocessor that suspends on specific commands based on a signal.
+
+    Parameters
+    ----------
+    signal: ``Signal``
+        The signal to subscribe to, whose value determines when to suspend.
+
+    commands: ``list of str``, optional
+        The commands to suspend on. If omitted, we'll suspend on all commands.
+
+    sleep: ``int`` or ``float``, optional
+        The amount of time to wait after `should_resume` returns ``True``
+        before ending the suspension. If `should_suspend` return ``True`` at
+        any time during this wait period, we will cancel the resumption and
+        return to the suspended state.
+    """
+    def __init__(self, signal, *, commands=None, sleep=0,
+                 pre_plan=null, post_plan=null, follow_plan=null):
+        self._sig = signal
+        self._cmd = commands
+        self._sleep = sleep
+        self._suspend_active = False
+        self._resume_ts = None
+        self._suspend_ev = Event()
+        self._ok_future = asyncio.Future()
+        self._ok_future.set_result('ok')
+        self._rlock = RLock()
+        self._subid = None
+
+    def should_suspend(self, value):
+        """
+        Returns ``True`` if we should suspend.
+
+        Parameters
+        ----------
+        value: signal value
+            The value reported by a signal callback.
+        """
+        raise NotImplementedError()
+
+    def should_resume(self, value):
+        """
+        Returns ``True`` if we should resume.
```

Comment:
I would note that the default behavior is just the reciprocal of the `should_suspend` unless implemented in child.

Reply:
Will-do